### ISSUE TITLE
Potential replacement with SAM2 backbone

### ISSUE THREAD

**Issue by TheMattBin** [heyoeyo/muggled_sam#22]
Hi, 

I'm working on replacing sam2 with [EfficientTAM](https://github.com/yformer/EfficientTAM/tree/main) since it claimed runs faster. However, I found the backbone used is quite different from sam2. It means I cannot replace easily and do the `state_dict` conversion.

It would be great if you can give some ideas on how to make the changes! Thanks!

**Comment by heyoeyo** [comment]
I'm not super familiar with EfficientTAM, but from looking at the repo, it seems like it's different from SAMv2 in two ways:
1. It uses a different image encoder
2. They've re-worked the way the memory attention/encoding is handled

As far as I can tell, they're using the image encoder from SAMv1 (see [ETAM](https://github.com/yformer/EfficientTAM/blob/abcd061ebd3cc6e7527d152d75b890126aaa53f6/efficient_track_anything/modeling/backbones/vitdet.py#L168-L170) vs. [SAMv1](https://github.com/facebookresearch/segment-anything/blob/dca509fe793f601edb92606367a655c15ac00fdf/segment_anything/modeling/image_encoder.py#L16) + [dectectron2](https://github.com/facebookresearch/detectron2/blob/a1ce2f956a1d2212ad672e3c47d53405c2fe4312/detectron2/modeling/backbone/vit.py#L234-L236)), except they aren't doing the [complicated relative position](https://github.com/heyoeyo/muggled_sam/tree/main/lib/v1_sam/components#decomposed-relative-position-encoder) encoding that SAMv1 used. They've also renamed some of the model components slightly. It might be possible to just copy the existing MugSAM [SAMv1 image encoder](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v1_sam/image_encoder_model.py#L26) and make (hopefully minor) modifications to support the ETAM model. Just at a glance, the differences I can spot are:
- The ETAM weights are all stored under a 'model' key. This is the same as SAMv2, but SAMv1 doesn't do this, so the v1 weight loading may need to be updated to [handle this properly](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v2_sam/make_sam_v2.py#L65-L67).
- The ETAM image encoder weights are mostly stored as `image_encoder.trunk...` whereas SAMv1 just uses `image_encoder...`. So the ETAM `state_dict` conversion would need to account for this. For example, the SAMv1 config code looks for [image_encoder.blocks](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v1_sam/state_dict_conversion/config_from_original_state_dict.py#L60), whereas the modified ETAM code would need to look for `image_encoder.trunk.blocks`. That being said, this is only needed if you want to dynamically figure out the ETAM model sizing (MugSAM does this to avoid needing hydra/.yaml config files). It would be a lot easier to [hard-code the config](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v1_sam/state_dict_conversion/config_from_original_state_dict.py#L27-L38) for ETAM to get things working faster.
- There are differences in the structure of the transformer `MLP` modules between [MSAM](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v1_sam/components/image_encoder_attention.py#L352-L356) vs. [SAMv1](https://github.com/facebookresearch/segment-anything/blob/dca509fe793f601edb92606367a655c15ac00fdf/segment_anything/modeling/common.py#L21-L22) vs. [ETAM](https://github.com/yformer/EfficientTAM/blob/abcd061ebd3cc6e7527d152d75b890126aaa53f6/efficient_track_anything/modeling/efficienttam_utils.py#L125-L127), specifically the way the layers are named/arranged. This will require changes to how the [state_dict conversion](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v1_sam/state_dict_conversion/convert_original_state_dict_keys.py#L179-L183) renames the ETAM weights to 'fit' into MugSAM. This may be a bit confusing, so what I'd recommend is to modify [this code](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v1_sam/state_dict_conversion/convert_original_state_dict_keys.py#L180-L183) to print out the changes:
```python
if "mlp.lin" in new_key:
  print("DEBUG STATE DICT CONVERSION")
  print("  ORIG KEY:", key)
  new_key = new_key.replace("lin1", "layers.0")
  new_key = new_key.replace("lin2", "layers.2")
  print("   NEW KEY:", new_key)
  return new_key
```
Then run MugSAM (e.g. use run_image.py or something) and load a SAMv1 model. The print out should show how the model weights are being renamed for MugSAM, and should make it easier to figure out what to do for handling ETAM (see the weight print-out code below for inspecting how the ETAM weights are named).
- Similarly, there are differences in the shape of the `neck` module between [MSAM](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v1_sam/image_encoder_model.py#L90-L95) vs. [SAMv1](https://github.com/facebookresearch/segment-anything/blob/dca509fe793f601edb92606367a655c15ac00fdf/segment_anything/modeling/image_encoder.py#L88-L104) vs. [ETAM](https://github.com/yformer/EfficientTAM/blob/abcd061ebd3cc6e7527d152d75b890126aaa53f6/efficient_track_anything/modeling/backbones/image_encoder.py#L68-L96). So again the existing [state_dict conversion](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v1_sam/state_dict_conversion/convert_original_state_dict_keys.py#L141-L143) will need to be updated to properly rename the ETAM weights to fit into MugSAM.
- As mentioned, ETAM isn't using relative position encodings in the transformer blocks. This may just mean that the relpos encoding can be deleted from the existing [SAMv1 implementation](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v1_sam/components/image_encoder_attention.py#L324) (along with removing the [associated components/weights](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v1_sam/components/image_encoder_attention.py#L287))
- There are some differences in the args used to initialize the ViT ([ETAM](https://github.com/yformer/EfficientTAM/blob/abcd061ebd3cc6e7527d152d75b890126aaa53f6/efficient_track_anything/modeling/backbones/vitdet.py#L174-L197) vs [SAMv1](https://github.com/facebookresearch/segment-anything/blob/dca509fe793f601edb92606367a655c15ac00fdf/segment_anything/modeling/image_encoder.py#L20-L35)). One obvious difference is `window_block_indexes` (ETAM) vs. `global_attn_indexes` (SAMv1). I assume these all end up working the same way, so I'm not sure if anything needs to be done, but it might be something to watch out for.

That's as much as I can tell from a quick look. For reference, I'm mostly basing this on looking at the model weight names of ETAM vs. SAMv1. For example, you can run this code to print out the image encoder weights for each model to compare them:
```python
# Print out image encoder weight names for loaded SAMv1/SAMv2/EfficientTAM model
model_path = "/path/to/model.pt"
state_dict = torch.load(model_path)
state_dict = state_dict["model"] if "model" in state_dict else state_dict
all_model_keys = tuple(state_dict.keys())
unique_sub_keys = {k.split(".")[0] for k in all_model_keys}
imgenc_keys = [k for k in all_model_keys if k.startswith("image")]
print(f"Image Encoder weights for {model_path}:", *imgenc_keys, sep="\n")
print("", "Unique sub keys:", *unique_sub_keys, sep="\n")
```
 Most of the weights look the same, but any time there are differences (e.g. the `neck` weights), it will require some changes to the loading used by MugSAM.

---

For the second point (differences in the memory attention/encoding). It's hard to see what's different (other than ETAM has [multiple RoPEAttention](https://github.com/yformer/EfficientTAM/blob/abcd061ebd3cc6e7527d152d75b890126aaa53f6/efficient_track_anything/modeling/memory_attention.py#L17-L19) implementations). What I'd probably recommend is to start by copying the MugSAM SAMv1 code over and modifying it to get ETAM working with image segmentation first. Once that's working, then it should be easier to copy over components one-by-one from SAMv2 and make the changes needed for supporting video segmentation.

It's a lot of finicky low-level changes, so if you have any questions or need clarification, let me know!

**Comment by heyoeyo** [comment]
> Should be not really impactful but for better results I think need to take care of those key not handled.

I think the first 3 (key not handled) aren't handled by MugSAMv2 either actually (the weights either aren't used due to default configs, or don't have much of an effect).
The last 2 (key missing) have to do with [samv2.0 vs. v2.1](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v2_sam/state_dict_conversion/config_from_original_state_dict.py#L325), so it might just mean that ETAM was based off of SAMv2.0 rather than v2.1.

> Also, I check original ETAM config for ViT patch size, I assume it's the same as SAMv1 config

The patch size is probably the same (or if it's different, the state_dict checks should figure it out automatically), but the [stride & padding](https://github.com/yformer/EfficientTAM/blob/abcd061ebd3cc6e7527d152d75b890126aaa53f6/efficient_track_anything/modeling/backbones/utils.py#L72-L73) values seem to follow the SAMv2 pattern (and these aren't detectable from weights), which could lead to noticeable changes in the output. Though ETAM may not use the default (4,4) & (3,3) values, I'm not sure.


> Does it mean that the model still output 3 features? Or it just my misunderstanding?

The use of high-res features shouldn't affect the number of features output by the model, though it does change whether the mask decoder tries to use 1 or 3 features from the image encoder.
The `multimask_output` setting is something separate. The SAM models always generate 4 mask predictions for any prompt (the 4 black & white masks visible on the right side of the [run_image.py script](https://github.com/heyoeyo/muggled_sam/blob/main/.readme_assets/run_image_anim.gif), for example). The `multimask_output` option is used in the original SAM code so that the model either returns only the [1st mask](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/modeling/sam/mask_decoder.py#L152) (if `multimask_output=False`) or the [last 3 masks](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/modeling/sam/mask_decoder.py#L147) (for `multimask_output=True`). Though with either setting, the model still generates 4 masks (MugSAM doesn't have this setting, since it always returns all 4 masks).

**Comment by heyoeyo** [comment]
> However, the runtime is somehow similar

With SAMv2, for tracking a single object, roughly 50% of the runtime is due to the image encoder and the other 50% is due to the memory attention step. The ETAM documentation seems to suggest that they changed the memory attention somehow, which I guess the MugSAMv2 code wouldn't account for, so that may be part of it.

Also, if you haven't tried it, encoding images at 512x512 (using `model.encode_image(img_bgr, max_side_length=512)`) gives a big speed up and some of the ETAM models seemed to be trained for this. If they used these models for the comparisons, it might explain some of the difference.

> May I know that if there's compile for ImageEncoder?

I did do some quick experiments with it, but there isn't anything in the MugSAM code. From what I remember, the components can be compiled the way that the [SAM2 code does it](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/sam2_video_predictor.py#L983). Though a lot of the components have errors when doing this, and need changes to the code to work properly.

Though the SAMv1 image encoder (and so probably ETAM also) is able to compile without changes I think. Something like this works:
```python
sammodel.image_encoder.forward = torch.compile(
  sammodel.image_encoder.forward
)
```
From what I remember, compiling tends to give roughly a 2x speed up. I don't think it works as-is for the memory fusion component unfortunately, which could make a big difference in video segmentation. I'm sure it can be made to work with some (hopefully minor) modifications, but the compiling errors aren't very clear, so I haven't taken the time to look through it. A lot of the SAMv2 code [mentions cloning](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/sam2_video_predictor.py#L1045-L1046) features to get compilation working, though I don't know why this is needed.

One other thing that might help is replacing the old [v1 attention](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v1_sam/components/image_encoder_attention.py#L294-L334) with a version that uses [scaled_dot_product_attention](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v2_sam/components/hiera_blocks.py#L270-L297), which seems to run more efficiently (v1 couldn't use this due to the [relative position encodings](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v1_sam/components/image_encoder_attention.py#L324), but I think ETAM doesn't use it).

**Comment by TheMattBin** [comment]
> From what I remember, the components can be compiled the way that the [SAM2 code does it](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/sam2_video_predictor.py#L983). Though a lot of the components have errors when doing this, and need changes to the code to work properly.

I also got quite a lot of errors as well when I try with the compiling methods in VOS, have no ideas on how to deal with it. So, I think just leave it there...

> The ETAM documentation seems to suggest that they changed the memory attention somehow, which I guess the MugSAMv2 code wouldn't account for, so that may be part of it.

I've checked scripts related memory attention (both ETAM and SAMv2), their scripts are almost the same except ETAM uses [EfficientRoPEAttention](https://github.com/yformer/EfficientTAM/blob/abcd061ebd3cc6e7527d152d75b890126aaa53f6/efficient_track_anything/modeling/sam/transformer.py#L317), I guess I can try to replace it with the existing RoPEAttention in MugSAMv2?

**Comment by heyoeyo** [comment]
> I also got quite a lot of errors as well when I try with the compiling methods in VOS, have no ideas on how to deal with it. So, I think just leave it there...

If you do want to try to get a little bit of extra performance out of the model, it is often possible to compile the sub-components without errors. For example, instead of doing: `model.memory_encoder.forward = torch.compile(...)`, which gives errors, you can do: `model.memory_encoder.sub_component.forward = torch.compile(...)` and that can sometimes work.

> I guess I can try to replace it with the existing RoPEAttention in MugSAMv2?

Ya if that's the only difference, then it should just be a matter of swapping that component out.

**Comment by TheMattBin** [comment]
I tried to use compile for different parts but I didn't manage to boost the speed, and sometimes get slower. So I guess the another method will be Onnx or Trt engine. I've gone through some onnx export repos and issues in MugSAM. The formers are similar to SAM2, just they use onnx version to run single object tracking, meaning quite a lot of efforts to hack c++ or python code to make as MOT. I wonder if you get any new idea on onnx now?

**Comment by heyoeyo** [comment]
> ... meaning quite a lot of efforts to hack c++ or python code to make as MOT

With the way SAMv2 works, I don't think it's possible to make a single onnx model that handles everything needed for object tracking, while being efficient/fast, and also handling MOT. The issue is that there is some amount of external state-management needed for tracking (which onnx doesn't seem to be well suited for), and additionally you want to cache/re-use some encodings for speed (which again, onnx doesn't really support).

However, it should be possible to break the model into multiple onnx models and then only need a small amount of additional code (e.g. python, c++ etc.) to manage moving data between the components. Supporting MOT should be fairly straight forward in this case as well.

>  I wonder if you get any new idea on onnx now?

I think the best approach would be to export each of the [major model components](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v2_sam/sam_v2_model.py#L47-L52) as their own onnx model (the `mask_decoder` and `memory_encoder` can be combined into 1 onnx model if you're only doing video segmentation). And then you'd just need to have a few functions outside of onnx (e.g. in python or c++) to handle (1) [initializing the tracking memory](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v2_sam/sam_v2_model.py#L239-L266) for an object, (2) [updating tracking](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v2_sam/sam_v2_model.py#L340-L382) on an object and potentially (3) [managing the memory bank](https://github.com/heyoeyo/muggled_sam/blob/7adadc2d2ea67c05d5677ab0d2e07f32305391f2/lib/v2_sam/memory_fusion_model.py#L190) for an object (if this is something that can't be converted to onnx, I'm not sure how well it supports dynamic data sizing operations like this).

This approach should work equally well for single or multi-object tracking. In this repo there's an [example script](https://github.com/heyoeyo/muggled_sam/blob/main/simple_examples/video_segmentation_multiobj.py) for MOT and the pattern should be very similar using onnx. Here's pseudo-code for how multi-object tracking could work with onnx models:

```python
for frame in video:
  # Encode image data (shared for all objects)
  encoded_img_tensors = onnx_img_encoder(frame)

  # Update tracking for each object independently for MOT
  # (assume we have 'obj_memory_data' storing previous mem encodings for objs)
  for obj_id in obj_ids_list:
    # Get mask/tracking prediction for object
    prev_obj_mem = obj_memory_data[obj_id]
    memfused_img_tensor = onnx_mem_fusion(encoded_img_tensors, prev_obj_mem)
    mask_pred_tensors = onnx_mask_decoder(memfused_img_tensor, encoded_img_tensors)

    # Store new memory encoding for object for use on next frame
    new_mem_tensor = onnx_mem_encoder(encoded_img_tensors, mask_pred_tensors)
    obj_memory_data[obj_id] = onnx_update_memory_bank(prev_obj_mem, new_mem_tensor)
```
This is leaving out some smaller details and is skipping the initialization step (i.e. creating the `obj_memory_data`), but that's basically the same code as above, except only running for the initial detection frame and without needing the `onnx_mem_fusion` step.
I think the only hard part, aside from getting the onnx export to work, is updating the memory bank since it follows some messy logic (though maybe there's a way to build this directly into an onnx model). If you don't need an exact 1-to-1 match with the original SAM2 outputs, you could throw away the use of `object pointers` to dramatically simplify the code (and onnx exporting), without much loss in tracking performance.
